CNS CSI Driver

divyenpatel opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:

This PR is adding a CNS CSI Driver and removing vSphere FCD CSI Driver.

Summary of change.
- Added a new controller for CNS CSI Driver at `pkg/csi/service/cns`
- Updated `pkg/csi/service/node.go`to add a logic for adding `AccessibleTopology` for CSINodes.
- Provided sample example files for CNS CSI Driver at `example`.
- Provided deployment YAML files for new driver at `manifests/1.14`
- Removed vSphere FCD Driver and related examples and deployment files.

**Special notes for your reviewer**:
- Name of the new CNS CSI driver is `csi.vsphere.vmware.com`
- End to End testing is in Progress.
- Full Sync/ Metadata Sync Container and E2E tests will be added in the next PRs.

**Integration Testing**

```
export VSPHERE_VCENTER=10.184.102.220
export VSPHERE_USER='Administrator@vsphere.local'
export VSPHERE_PASSWORD='Admin!23'
export VSPHERE_DATASTORE_URL='ds:///vmfs/volumes/vsan:52b5fd9468ec70b4-5f6df02a13ac56d1/'
export VSPHERE_K8S_NODE='k8s-node1'
export KUBECONFIG='/Users/divyenp/.kube/config'
export VSPHERE_INSECURE=1
export VSPHERE_DATACENTER='vcqaDC'
export VSPHERE_STORAGE_POLICY_NAME='vSAN Default Storage Policy'


$ make integration-unit-test
go test -v -count=1 -tags=integration-unit ./pkg/csi/service/cns
=== RUN   TestCreateVolumeWithStoragePolicy
--- PASS: TestCreateVolumeWithStoragePolicy (13.15s)
=== RUN   TestCompleteControllerFlow
--- PASS: TestCompleteControllerFlow (13.76s)
    controller_test.go:467: ControllerPublishVolume will be called with req {VolumeId:d676fab5-2297-4503-b5dc-c5ec278cb5fa NodeId:k8s-node1 VolumeCapability:access_mode:<mode:SINGLE_NODE_WRITER >  Readonly:false Secrets:map[] VolumeContext:map[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
    controller_test.go:473: ControllerPublishVolume succeed, diskUUID 6000c29c00690c5a2a46d89e0057bd61 is returned
    controller_test.go:480: ControllerUnpublishVolume will be called with req {VolumeId:d676fab5-2297-4503-b5dc-c5ec278cb5fa NodeId:k8s-node1 Secrets:map[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
    controller_test.go:485: ControllerUnpublishVolume succeed
PASS
ok      sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/cns      27.246s

```

**Unit Testing**

```
$ make unit-test
env -u VSPHERE_VCENTER -u VSPHERE_DATACENTER -u VSPHERE_PASSWORD -u VSPHERE_USER -u VSPHERE_DATASTORE_URL -u VSPHERE_STORAGE_POLICY_NAME -u VSPHERE_K8S_NODE -u VSPHERE_INSECURE -u KUBECONFIG go test -v -count=1 ./pkg/csi/service ./pkg/csi/service/cns ./pkg/csi/testing
=== RUN   TestGetDisk
=== RUN   TestGetDisk/#00
=== PAUSE TestGetDisk/#00
=== RUN   TestGetDisk/#01
=== PAUSE TestGetDisk/#01
=== CONT  TestGetDisk/#00
=== CONT  TestGetDisk/#01
--- PASS: TestGetDisk (0.00s)
    --- PASS: TestGetDisk/#00 (0.00s)
    --- PASS: TestGetDisk/#01 (0.00s)
PASS
ok  	sigs.k8s.io/vsphere-csi-driver/pkg/csi/service	0.386s
=== RUN   TestCreateVolumeWithStoragePolicy
E0904 15:06:31.647318   46141 config.go:202] No Virtual Center hosts defined
--- PASS: TestCreateVolumeWithStoragePolicy (0.04s)
=== RUN   TestCompleteControllerFlow
--- PASS: TestCompleteControllerFlow (0.01s)
    controller_test.go:467: ControllerPublishVolume will be called with req {VolumeId:0306e12b-d86e-43bf-9597-f5f3463b2434 NodeId:DC0_C0_RP0_VM0 VolumeCapability:access_mode:<mode:SINGLE_NODE_WRITER >  Readonly:false Secrets:map[] VolumeContext:map[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
    controller_test.go:473: ControllerPublishVolume succeed, diskUUID 6000c298595bf4575739e9105b2c0c2d is returned
    controller_test.go:480: ControllerUnpublishVolume will be called with req {VolumeId:0306e12b-d86e-43bf-9597-f5f3463b2434 NodeId:DC0_C0_RP0_VM0 Secrets:map[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
    controller_test.go:485: ControllerUnpublishVolume succeed
PASS
ok  	sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/cns	0.416s
=== RUN   TestCSI
Running Suite: CSI Suite
========================
Random Seed: 1567634791
Will run 3 of 4 specs

time="2019-09-04T15:06:31-07:00" level=info msg="identity service registered"
time="2019-09-04T15:06:31-07:00" level=info msg="controller service registered"
time="2019-09-04T15:06:31-07:00" level=info msg="node service registered"
time="2019-09-04T15:06:31-07:00" level=info msg=serving endpoint="memu://csi-vsphere-test"
time="2019-09-04T15:06:31-07:00" level=info msg="gracefully stopped"
•time="2019-09-04T15:06:31-07:00" level=info msg="identity service registered"
time="2019-09-04T15:06:31-07:00" level=info msg="controller service registered"
time="2019-09-04T15:06:31-07:00" level=info msg="node service registered"
time="2019-09-04T15:06:31-07:00" level=info msg=serving endpoint="memu://csi-vsphere-test"
time="2019-09-04T15:06:31-07:00" level=info msg="gracefully stopped"
•
------------------------------
P [PENDING]
CSI plugin
/Users/divyenp/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/pkg/csi/testing/plugin_startup_test.go:35
  Started with memory pipe
  /Users/divyenp/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/pkg/csi/testing/plugin_startup_test.go:53
    when Plugin unconfigured
    /Users/divyenp/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/pkg/csi/testing/plugin_startup_test.go:58
      Identity Service
      /Users/divyenp/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/pkg/csi/testing/plugin_startup_test.go:67
        Probe
        /Users/divyenp/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/pkg/csi/testing/plugin_startup_test.go:100
          Should fail
          /Users/divyenp/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/pkg/csi/testing/plugin_startup_test.go:101
------------------------------
time="2019-09-04T15:06:31-07:00" level=info msg="identity service registered"
time="2019-09-04T15:06:31-07:00" level=info msg="controller service registered"
time="2019-09-04T15:06:31-07:00" level=info msg="node service registered"
time="2019-09-04T15:06:31-07:00" level=info msg=serving endpoint="memu://csi-vsphere-test"
time="2019-09-04T15:06:31-07:00" level=info msg="gracefully stopped"
•
Ran 3 of 4 Specs in 0.004 seconds
SUCCESS! -- 3 Passed | 0 Failed | 1 Pending | 0 Skipped
--- PASS: TestCSI (0.00s)
PASS
ok  	sigs.k8s.io/vsphere-csi-driver/pkg/csi/testing	0.406s
```


**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Added CNS CSI Driver and removed FCD CSI Driver
```

divyenpatel commented:
/assign @codenrhoden 
/assign @frapposelli
/assign @dvonthenen
/assign @andrewsykim

codenrhoden commented:
@divyenpatel I'll take a deeper look at this tomorrow morning.

Don't worry about anything from `staticcheck`. That's something new, and it's currently set to `optional`, meaning that it's not required to pass that test to merge. Looks like something weird is happening with it. :/

One comment based solely on the description of this PR, though, is that I thought the agreement had been not to remove the FCD driver until all the CNS functionality was merged in.

divyenpatel commented:
Thanks @codenrhoden , I see I was attempting to fix `verify-staticcheck`. It is passing locally though

```
$ ./hack/check-staticcheck.sh 
go: finding honnef.co/go/tools/cmd/staticcheck latest
go: finding honnef.co/go/tools/cmd latest
```

Regarding the removal of FCD driver, I initially started with keeping both drivers and used a flag to determine which controller to initialize, but later realized rest of the components full sync and metadata syncer are completely independent of the controller and node plugin, so removed it along with this PR.

codenrhoden commented:
> Thanks @codenrhoden , I see I was attempting to fix `verify-staticcheck`. It is passing locally though
> 
> ```
> $ ./hack/check-staticcheck.sh 
> go: finding honnef.co/go/tools/cmd/staticcheck latest
> go: finding honnef.co/go/tools/cmd latest
> ```
> 
Great! It had dawned on my why it is failing in CI, but not locally. I can fix that in the morning as well. (After the tool is installed, it's location isn't in $PATH -- easy fix).

codenrhoden commented:
https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/53 fill fix the failing tests

dvonthenen commented:
/retest

dvonthenen commented on pkg/csi/service/cns/controller.go:
```diff
@@ -0,0 +1,392 @@
+	*csi.ListVolumesResponse, error) {
+
+	klog.V(4).Infof("ListVolumes: called with args %+v", *req)
+	return nil, status.Error(codes.Unimplemented, "")
```
You return Unimplemented here, but in the list capabilities, you have RPC_LIST_VOLUMES present. You might want to remove RPC_LIST_VOLUMES since it could potentially screw up other software that will depend on that actually working. As example of such a program, something like Velero.

divyenpatel replied:
removed `csi.ControllerServiceCapability_RPC_LIST_VOLUMES` from `ControllerGetCapabilities`.

dvonthenen commented on pkg/csi/service/node_test.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2018 The Kubernetes Authors.
+Copyright 2019 The Kubernetes Authors.
```
Don't really need to update since the file exists, but it's all good.

codenrhoden replied:
No it does matter. This shouldn't be changed. :)

divyenpatel replied:
done reverted Copyright year.

codenrhoden commented on pkg/csi/service/cns/nodes.go:
```diff
@@ -0,0 +1,247 @@
+	"fmt"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"golang.org/x/net/context"
```
just "context"

divyenpatel replied:
fixed.

codenrhoden commented on pkg/csi/service/node.go:
```diff
@@ -362,23 +353,90 @@ func (s *service) NodeGetInfo(
+	var accessibleTopology map[string]string
+	topology := &csi.Topology{}
+
+	if cfg.Labels.Zone != "" && cfg.Labels.Region != "" {
```
Okay, so to answer my question from above the nodes need access to vCenter in order to figure out topology. I guess there really is no way around that...

codenrhoden commented on pkg/csi/types/types.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2018 The Kubernetes Authors.
+Copyright 2019 The Kubernetes Authors.
```
don't change this.

divyenpatel replied:
reverted to 2018.

codenrhoden commented on manifests/1.14/rbac/vsphere-csi-controller-rbac.yaml:
```diff
@@ -15,6 +15,24 @@ rules:
     resources: ["csidrivers"]
     verbs: ["create", "delete"]
+  - apiGroups: [""]
+    resources: ["nodes"]
```
@andrewsykim Would love to get your eyes on the RBAC roles here...

divyenpatel replied:
@codenrhoden @andrewsykim 

moved them up. 

I have also removed RBAC rules for `secrets` and `csidrivers`. so revised rules will be as below.

```
rules:
  - apiGroups: [""]
      resources: ["nodes"]
      verbs: ["get", "list", "watch"]
  - apiGroups: [""]
    resources: ["persistentvolumes"]
    verbs: ["get", "list", "watch", "update", "create", "delete"]
  - apiGroups: [""]
    resources: ["persistentvolumeclaims"]
    verbs: ["get", "list", "watch", "update"]
  - apiGroups: [""]
    resources: ["pods"]
    verbs: ["get", "list", "watch"]
  - apiGroups: ["storage.k8s.io"]
    resources: ["csinodes"]
    verbs: ["get", "list", "watch"]
  - apiGroups: ["storage.k8s.io"]
    resources: ["volumeattachments"]
    verbs: ["get", "list", "watch", "update"]
```

rules for `persistentvolumes`, `persistentvolumeclaims` and `pods` are needed for metadata syncer and full sync containers, rest all are used by controller, provisioner and attacher.

codenrhoden replied:
I don't think I was clear enough on what I meant, here...

Instead of listing all these out separately, like this:

```yaml
  - apiGroups: [""]
    resources: ["secrets"]
    verbs: ["get", "list", "watch"]
  - apiGroups: [""]
    resources: ["nodes"]
    verbs: ["get", "list", "watch"]
  - apiGroups: [""]
    resources: ["persistentvolumeclaims"]
    verbs: ["get", "list", "watch"]
  - apiGroups: [""]
    resources: ["pods"]
    verbs: ["get", "list", "watch"]
```

All four of those can be combined into one:

```yaml
  - apiGroups: [""]
    resources: ["secrets", "nodes", "persistentvolumeclaims", "pods"]
    verbs: ["get", "list", "watch"]
```

Since the `apiGroup` and `verbs` are the same, you can list multiple `resources`.

Also, in the current iteration of this, it seems `secrets` have been deleted entirely. I don't think that's what you wanted.

dvonthenen commented on manifests/1.14/deploy/vsphere-csi-node-ds.yaml:
```diff
@@ -38,22 +40,30 @@ spec:
               value: "false"
-#           - name: X_CSI_DEBUG
-#             value: "true"
+            - name: X_CSI_VSPHERE_CLOUD_CONFIG
```
Yea, I that would not be a very good situation to be in. You have the potential for security-related issues. Originally creds were only exposed to just the master node which you would argue is held to a higher standard since you're talking control plane at that point.

divyenpatel replied:
@codenrhoden @dvonthenen 

mounting  `vsphere-config-volume` config-map on vsphere-csi-node pods is optional.

node plugin only needs access to the `csi-vsphere.conf`, when zone feature is enabled, so it can obtain and return accessibleTopology of the node from `NodeGetInfo`.

If we do not have zone/region configured on the cluster then we do not need to mount this credential file on `vsphere-csi-node` pods.

divyenpatel commented on pkg/csi/service/cns/controller.go:
```diff
@@ -0,0 +1,392 @@
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	cnstypes "github.com/vmware/govmomi/cns/types"
+	"github.com/vmware/govmomi/units"
+	"golang.org/x/net/context"
```
done. imported `context ` from stdlib.
replaced this at 5 other places in this repository.

divyenpatel commented on manifests/1.14/deploy/vsphere-csi-controller-ss.yaml:
```diff
@@ -15,27 +17,43 @@ spec:
+              name: socket-dir
         - name: vsphere-csi-controller
-          image: gcr.io/cloud-provider-vsphere/csi/release/driver:v0.2.1
+          image: cloudnativestorage/vsphere-block-csi-driver:latest
```
Thanks for catching this.

We have released GA images at `vmware/vsphere-block-csi-driver:v1.0.0` and `vmware/volume-metadata-syncer:v1.0.0`.  Let me replace them for now, to keep in-sync with official documentation. - https://docs.vmware.com/en/VMware-vSphere/6.7/Cloud-Native-Storage/GUID-039425C1-597F-46FF-8BAA-C5A46FF10E63.html

We can change the image location to `gcr.io/cloud-provider-vsphere/csi/release/driver:<new-tag>`, when new images are available from this location after we cut a new tag release on `vsphere-csi-driver` repository.

We can then change official vSphere documentation to use YAML files directly from this repository with github URL, instead of asking user to copy/paste from documentation.

divyenpatel commented on example/vsphere.conf:
```diff
@@ -0,0 +1,14 @@
+[VirtualCenter "1.2.3.4"]
+insecure-flag = "true"
+user = "vCenter username"
+password = "vCenter password"
```
Yes agree. Let me remote this file from example. 

We have already documented steps at https://docs.vmware.com/en/VMware-vSphere/6.7/Cloud-Native-Storage/GUID-039425C1-597F-46FF-8BAA-C5A46FF10E63.html

We will also make sure that mark down documentation files in this repository describes all available options.

andrewsykim commented on pkg/csi/types/constants.go:
```diff
@@ -14,16 +14,11 @@ See the License for the specific language governing permissions and
-var (
-	ErrListInvalidNextToken = errors.New(ListInvalidNextTokenErrMsg)
+	// LabelRegionFailureDomain is label placed on nodes and PV containing region detail
+	LabelRegionFailureDomain = "failure-domain.beta.kubernetes.io/region"
```
FYI for a future PR, these are also defined in k8s.io/api/core/v1 for external consumers

codenrhoden replied:
https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/55

codenrhoden commented on manifests/1.14/rbac/attacher-rbac.yaml:
```diff
@@ -1,39 +0,0 @@
-apiVersion: v1
```
One more question I had... Did you run into problems launching the attacher and provisioner separately? They were split out with our 1.14 manifests so that they could have different service accounts and RBAC roles attached them. The CSI driver itself specifically had very limited permissions to do anything with the K8s API. Looks like all the permissions for everything have been re-combined into one big role, which to me violates the idea of least-based privilege.

divyenpatel replied:
We started with creating 4 service accounts 1 for each pod attacher, provisioner, controller, and metadata syncer.

Later while deploying them on the setup having multiple master nodes, we realized all these pods are getting deployed randomly on multiple master nodes. 

We want to make sure attacher, provisioner and controller remains in a single node, as they communicate over Unix domain socket. If they are on different nodes, the system does not work.

so to ensure guarantee, these containers run on a single node, we embedded all of them in a single pod.

Only one service account can be assigned to a pod, so we merged all roles created for these individual service accounts and assigned them to a single service account `vsphere-csi-controller`.

divyenpatel commented on manifests/1.14/deploy/vsphere-csi-controller-ss.yaml:
```diff
@@ -52,19 +70,47 @@ spec:
+          env:
+            - name: X_CSI_FULL_SYNC_INTERVAL_MINUTES
+              value: "30"
+            - name: X_CSI_VSPHERE_CLOUD_CONFIG
```
Yes, let me remove `X_CSI` prefix from env variables not related CSI.

divyenpatel commented on manifests/1.14/deploy/vsphere-csi-node-ds.yaml:
```diff
@@ -38,22 +40,30 @@ spec:
-#             value: "true"
+            - name: X_CSI_VSPHERE_CLOUD_CONFIG
+              value: "/etc/cloud/csi-vsphere.conf" # here csi-vsphere.conf is the name of the file used for creating secret using "--from-file" flag
+            - name: X_CSI_DEBUG
```
Yes let us be consistent. I am removing `X_CSI_DEBUG ` flag from here.

divyenpatel commented on manifests/1.14/rbac/vsphere-csi-controller-rbac.yaml:
```diff
@@ -15,6 +15,24 @@ rules:
+    verbs: ["get", "list", "watch", "update"]
+  - apiGroups: [""]
+    resources: ["persistentvolumeclaims"]
+    verbs: ["get", "list", "watch", "update"]
```
metadata syner and full sync requires get, list and watch on `persistentvolumeclaims ` and `persistentvolumes `.

Let me check if update is needed. if not I will remove it.

divyenpatel replied:
we do not need to update PV, PVC or POD spec , so cleaning that up as well.

divyenpatel replied:
Update on PV is required by provisioner to add finalizer, but update on PVC is not needed so removed that access rule from `persistentvolumeclaims `.

divyenpatel replied:
regarding creating different account/role. please refer - https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/52#discussion_r319694600

divyenpatel commented:
@codenrhoden @andrewsykim @dvonthenen addressed all review comments here - https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/52/commits/6915b99d27248ec335f9f4acfc4f2dc94b6b66cc

divyenpatel commented:
@codenrhoden 

I have verified end to end workflow, with this change by creating storageclass, pvc, pod. and also deleting these resources. no issue observed.

only issue I am trying to debug is why I am not able to see the logs for controller pod. For that I am trying to add `klog.InitFlags(nil)` in `cmd/vsphere-csi/main.go` as below.

```golang
// main is ignored when this package is built as a go plug-in.
func main() {
	klog.InitFlags(nil)
	flag.Parse()
	gocsi.Run(
		context.Background(),
		service.Name,
		"A CSI plugin for VMware vSphere storage",
		usage,
		provider.New())
}
```

but this is failing with 

```
panic: ./main flag redefined: log_dir

goroutine 1 [running]:
flag.(*FlagSet).Var(0xc0000ae0c0, 0x2a55700, 0x3c12550, 0x276d441, 0x7, 0x27c9d17, 0x2f)
        /usr/local/Cellar/go/1.12.1/libexec/src/flag/flag.go:850 +0x4af
flag.(*FlagSet).StringVar(...)
        /usr/local/Cellar/go/1.12.1/libexec/src/flag/flag.go:753
k8s.io/klog.InitFlags(0x0)
        /Users/divyenp/go/pkg/mod/k8s.io/klog@v0.2.0/klog.go:411 +0xa4
main.main()
        /Users/divyenp/go/src/github.com/kubernetes-sigs/vsphere-csi-driver/cmd/vsphere-csi/main.go:31 +0x2e
```

seems like`log_dir` flag is already getting defined by some dependency.

divyenpatel commented:
@codenrhoden we need to delete `secrets` entirely, as we are not listening on any changes on the secret. In CSI driver, we are directly mounting secret in the container as volume, so when credential is changed, and prior credentials are no longer valid we re-read new credentials from the config file and re-establishing sessions.

dvonthenen commented:
> hmm, perhaps @dvonthenen has some insight on what might be happening with the `klog` `log_dir` stuff?

I would definitely take a look at the existing implementation in this repo for guidance. That's done here:
https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/csi/service/service.go#L119-L135

You can also find examples for initializing klog here:
https://github.com/kubernetes/klog/tree/master/examples

dvonthenen commented:
> @codenrhoden we need to delete `secrets` entirely, as we are not listening on any changes on the secret. In CSI driver, we are directly mounting secret in the container as volume, so when credential is changed, and prior credentials are no longer valid we re-read new credentials from the config file and re-establishing sessions.

I would definitely change this when you get a chance as that can be problematic from my experience.

divyenpatel commented:
@dvonthenen `klog initialization` issue is resolved with fixing go.mod dependecy as part of this commit - https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/52/commits/7e80bb194700d67443ce018fd9f973a92e1b8d5a


Locally make check is also passing. 

```

$ make check
hack/check-format.sh
go: finding golang.org/x/tools/cmd/goimports latest
go: finding golang.org/x/tools/cmd latest
go: finding golang.org/x/tools latest
hack/check-lint.sh
go: finding golang.org/x/lint/golint latest
go: finding golang.org/x/lint latest
hack/check-mdlint.sh
hack/check-shell.sh
hack/check-staticcheck.sh
go: finding honnef.co/go/tools/cmd/staticcheck latest
go: finding honnef.co/go/tools/cmd latest
hack/check-vet.sh
``` 

Let me see what broke the kubernetes prow robot.

codenrhoden commented:
@divyenpatel I'm not sure what it doesn't like about that commit message. Maybe it's a hyperlink?

I've only seen that message when someone puts the GitHub magic syntax in there for autoclosing an issue (e.g. `fixes #xxx`). I know that's the main one they are trying to avoid.

Anyways, once you are happy with everything, squash it all down to one I'd say and then you won't have the issue anymore. I'll look through things one more time, but I'd say this should get merged today. Thanks for all your responsiveness!

codenrhoden commented:
> @divyenpatel I'm not sure what it doesn't like about that commit message. Maybe it's a hyperlink?
> 
> I've only seen that message when someone puts the GitHub magic syntax in there for autoclosing an issue (e.g. `fixes #xxx`). I know that's the main one they are trying to avoid.

Ah, looking at the source code (https://github.com/kubernetes/test-infra/blob/8ac75e2cba385e9dfb96157da756c56d3ac677c7/prow/plugins/invalidcommitmsg/invalidcommitmsg.go#L52-L53), it's the `@` symbol in the hyperlink... That's a pretty heavy hammer.

divyenpatel commented:
@codenrhoden I have squashed all my commits. Please take a look again.

Also adding integration and unit-testing result on the PR description.

codenrhoden commented on pkg/csi/types/constants.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2018 The Kubernetes Authors.
+Copyright 2019 The Kubernetes Authors.
```
Since this file is just renamed/moved, the copyright year should stay at 2018.

divyenpatel replied:
done. fixed.

codenrhoden commented:
/lgtm
/approve

DaniDD commented:
I need the FCD driver. Wo dont have vSAN, only FCD. Can you please reintegrate it?
